arch/src/aarch64/fdt.rs

```diff
@@ -0,0 +1,424 @@
+        -> c_int;
+    fn fdt_end_node(fdt: *mut c_void) -> c_int;
+    fn fdt_open_into(fdt: *const c_void, buf: *mut c_void, bufsize: c_int) -> c_int;
+    fn fdt_finish(fdt: *const c_void) -> c_int;
+    fn fdt_pack(fdt: *mut c_void) -> c_int;
+}
+
+#[derive(Debug)]
+pub enum Error {
+    AppendFDTNode(io::Error),
+    AppendFDTProperty(io::Error),
+    CreateFDT(io::Error),
+    CstringFDTTransform(NulError),
+    FinishFDTReserveMap(io::Error),
+    IncompleteFDTMemoryWrite,
+    WriteFDTToMemory(GuestMemoryError),
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+// Creates the flattened device tree for this VM.
+pub fn create_fdt(guest_mem: &GuestMemory, num_cpus: u32, cmdline: &CStr) -> Result<(Vec<u8>)> {
+    // Alocate stuff necessary for the holding the blob.
+    let mut fdt = vec![0; FDT_MAX_SIZE];
+    // Safe since we allocated this array with FDT_MAX_SIZE.
+    let mut fdt_ret = unsafe { fdt_create(fdt.as_mut_ptr() as *mut c_void, FDT_MAX_SIZE as c_int) };
+
+    if fdt_ret != 0 {
+        return Err(Error::CreateFDT(io::Error::last_os_error()));
+    }
+
+    // The flattened device trees created with fdt_create() contains a list of
+    //reserved memory areas. Wneed to call fdt_finish_reservemap so as to make sure that there is a
+    //terminator in the reservemap list and whatever happened to be at the
+    //start of the FDT data section would end up being interpreted as
+    //reservemap entries.
+    // Safe since we previously allocated this array.
+    fdt_ret = unsafe { fdt_finish_reservemap(fdt.as_mut_ptr() as *mut c_void) };
+    if fdt_ret != 0 {
+        return Err(Error::FinishFDTReserveMap(io::Error::last_os_error()));
+    }
+
+    // For an explanation why these nodes were introduced in the blob take a look at
+    // https://github.com/torvalds/linux/blob/master/Documentation/devicetree/booting-without-of.txt#L845
+    // Look for "Required nodes and properties".
+
+    // Header or the root node as per above mentioned documentation.
+    begin_node(&mut fdt, "")?;
+    property_string(&mut fdt, "compatible", "linux,dummy-virt")?;
+    // For info on #address-cells and size-cells read "Note about cells and address representation"
+    // from the above mentioned txt file.
+    property_u32(&mut fdt, "#address-cells", ADDRESS_CELLS)?;
+    property_u32(&mut fdt, "#size-cells", SIZE_CELLS)?;
+    // This is not mandatory but we use it to point the root node to the node
+    // containing description of the interrupt controller for this VM.
+    property_u32(&mut fdt, "interrupt-parent", GIC_PHANDLE)?;
+    create_cpu_nodes(&mut fdt, num_cpus)?;
+    create_memory_node(&mut fdt, guest_mem)?;
+    create_chosen_node(&mut fdt, cmdline)?;
+    create_gic_node(&mut fdt, u64::from(num_cpus))?;
+    create_timer_node(&mut fdt)?;
+    create_psci_node(&mut fdt)?;
+
+    // End Header node.
+    end_node(&mut fdt)?;
+
+    let fdt_address = GuestAddress(get_fdt_addr(&guest_mem));
```

Comment:
All the code that follows seems disconnected from the actual logic of creating a FDT. Would it make sens to move it to another function ? If you decide to use a builder it can be moved to a method called `finish()` or `commit()`

Reply:
Done. I moved the code to another function.